source/slang/parser.cpp

```diff
@@ -492,34 +537,36 @@ namespace Slang
 
     Token Parser::ReadToken(TokenType expected)
     {
-        if (tokenReader.PeekTokenType() == expected)
+        if (PeekTokenType() == expected)
         {
             isRecovering = false;
-            return tokenReader.AdvanceToken();
+            return AdvanceToken();
         }
 
         if (!isRecovering)
         {
             Unexpected(this, expected);
-            return tokenReader.PeekToken();
+            return PeekToken();
         }
         else
         {
             // Try to find a place to recover
             if (TryRecoverBefore(this, expected))
             {
                 isRecovering = false;
-                return tokenReader.AdvanceToken();
+                return AdvanceToken();
             }
 
-            return tokenReader.PeekToken();
+            return PeekToken();
         }
     }
 
     bool Parser::LookAheadToken(const char * string, int offset)
     {
+        if (offset < genericCloserCount)
```

Comment:
This ends up being a bit messy, and probably warrants commenting to make sure a reader would understand how the `offset` applies to the conceptual lookahead.